[dohaduong] iP

dohaduong opened the pull request:
### Duke 
**Your friendly task management app**

> "Until we can manage time, we can manage nothing else." - Peter Drucker [source](https://www.tameday.com/motivational-quotes-for-work-productivity/)

Your busy life deserves this.
**Duke** is 

- an _efficient_ task management app 💯 
- _easy_ to use 👍 
- helps you stay _organized_ 🥇 

To start using Duke, you only need to follow these **steps**: 

1. Download and install Duke
2. Add your tasks and deadlines
3. Let Duke handle your timetable for you!

**Available features:**

- [x] Managing tasks
- [ ] Managing deadlines and events
- [ ] Reminder for deadlines and events

Duke can be used to practice your Java as a programmer too! Here's the `main` method for you to start with: 
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

dohaduong committed:
Level 1 completed

dohaduong committed:
Level 1 completed

dohaduong committed:
Level 1 completed

dohaduong committed:
Level 2 completed

dohaduong committed:
.\Duke.Level 3 completed

dohaduong committed:
Level 4 completed

dohaduong committed:
test case

dohaduong committed:
test case

dohaduong committed:
Level 5 completed

dohaduong committed:
Level 6 completed

dohaduong committed:
Enums already used in previous levels

dohaduong committed:
Level 7 file saver completed

dohaduong committed:
Level 7 file saver completed

dohaduong committed:
edit

dohaduong committed:
Level 7 file saver completed

dohaduong committed:
Level 8 Date and Time

dohaduong committed:
Merge level 7 and 8 to master

dohaduong committed:
Merge level 7 and 8

dohaduong committed:
More OOP

dohaduong committed:
divided into packages

dohaduong committed:
Class divided into packages

dohaduong committed:
Merge branch 'add-gradle-support'

dohaduong committed:
JUnit tests

dohaduong committed:
jar file created - ip main

dohaduong committed:
JavaDoc added for public methods

dohaduong committed:
Coding Standard followed

dohaduong committed:
Level 9 completed

dohaduong committed:
Merge A-CodingStandard

dohaduong committed:
merge

dohaduong committed:
Merge Level-9

dohaduong committed:
Level 9 completed and merged

dohaduong committed:
Level 9 completed and edited

dohaduong committed:
Handle conflicts

dohaduong committed:
Resolve conflicts

pkpaing commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,199 @@
+import java.io.IOException;
+
+
+enum action {
+    Bye,
+    List,
+    Mark,
+    Unmark,
+    Todo,
+    Deadline,
+    Event,
+    Delete,
+    Find
+}
```
Nice use of enum! Class/enum names are recommended to be nouns and written in PascalCase (e.g. Action instead of action)

pkpaing commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,170 @@
+     * @return the String detail.
+     * @throws duke.MissingContentException if arr is empty.
+     */
+    public String deadlineDetail(String[] arr) throws MissingContentException {
```
Good use of camelCase for method names!

pkpaing reviewed (commented):
Looks good!

dohaduong committed:
Checkstyles added

dohaduong committed:
GUI

dohaduong committed:
Level 10 in progress

dohaduong committed:
Level 10 completed

dohaduong committed:
Level 10

dohaduong committed:
Level 10 updated

dohaduong committed:
Level 10 done

jugheadjones10 commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,258 @@
+     * @param arr array of original array
+     * @return full string command for users
+     */
+    public static String toDo(String[] arr) {
```
I think it might be more suitable if you name the method as a **verb** instead of a **noun** like "toDo". Maybe something like parseToDo? This way I will immediately know what the method does when I see the method name.

jugheadjones10 commented on src/main/java/duke/command/TaskList.java:
```diff
@@ -0,0 +1,242 @@
+     * @throws IOException if array does not exist
+     */
+    public String[] readTaskList() throws IOException {
+        try {
+            try {
+                return this.arr;
+            } catch (IndexOutOfBoundsException e) {
+                throw new MissingContentException();
+            }
+        } catch (MissingContentException e) {
+            System.out.println(e.getMessage());
+        }
+        return this.arr;
```
I'm not entirely sure whether using nested try catch statements makes sense. An alternative would be using just one try catch statement, but in the catch statement you check for different types of errors and handle differently from there.

jugheadjones10 commented on src/main/java/duke/command/TaskList.java:
```diff
@@ -0,0 +1,242 @@
+     * @return new task list with task unmarked.
+     * @throw InvalidIndexException if array at specific index is null
+     */
+    public String unmark(int num1) throws IOException {
```
Any particular reason you chose to use "num1" as the argument name? Seems a bit inconsistent with the other methods that simply use "num". Also, having a number in a variable name might be a bit unreadable.

jugheadjones10 reviewed (commented):
Great work! Some minor readability comments.

dohaduong committed:
Create a new JAR file for Duke

Update Duke file with better UI and thus better users' experience.

Let's use Gradle to create the new JAR file for Duke.

Using Gradle is more preferable as it is easier and quicker.

dohaduong committed:
Add use of assertions for some public and private methods.
>>
>> Ui and Parser classes are responsible for interacting with users and users' input handling
>> and thus, there could be many errors.
>>
>> Use of assertion was added to these classes.
>>
>> This is to make sure the the programmer awares of the error.

dohaduong committed:
Merge pull request #1 from dohaduong/branch-A-Assertions

Add use of assertions for some public and private methods.

dohaduong committed:
Improve Code Quality.

Code Quality is highly needed for understandability.

Critically examines the code and refactor to improve the code quality where necessary.
Magic numbers were added and long methods with deep loops are avoided.

This is to make sure the code is easy to understand.

dohaduong committed:
Merge pull request #2 from dohaduong/branch-A-CodeQuality

Create PR from branch-A-CodeQuality

dohaduong committed:
CI setup completed.

dohaduong committed:
Added new extension: Undo

Users may want to undo the very last command.

Undo command has been added to Duke.
If the last command is only for listing/finding purpose, it will not be able to undone as these commands do not change any data that Duke has.

Users can now undo the last command, given that it is applicable.

dohaduong committed:
Add new extension: Undo

Users may want to undo the very last command

let's,
* update Duke's available commands so that undo can become a legit command
* update Parser and Ui so that Duke can revert back to before the most recent command

We need to update Parser and Ui classes as these classes are used to
reply to users and get input from users directly

joytqt-1202 commented on src/main/java/duke/command/TaskList.java:
```diff
@@ -0,0 +1,250 @@
+                            + String.format("Now you have %d "
+                            + "tasks in the list", this.getValidLen()));
+                }
+            } catch (IndexOutOfBoundsException e) {
+                throw new InvalidIndexException();
+            }
+        } catch (InvalidIndexException e) {
+            return (e.getMessage());
+        }
+        return new InvalidIndexException().getMessage();
```
good catching of exception cases! 
but instead of nested `try-catch` blocks, could you try to use `try-catch-catch` instead?
`try (...) { ... } catch (...) { ... } catch (...) { ... }`

dohaduong committed:
Minor changes to improve code quality

dohaduong committed:
Minor changes to improve Code Quality

dohaduong committed:
Minor edit to improve code quality

dohaduong committed:
Improve code quality

dohaduong committed:
Update GUI: Avatar shaped to circle

dohaduong committed:
Update GUI: Avatar in circle

dohaduong committed:
Merge pull request #4 from dohaduong/A-BetterGui

Update GUI: Avatar in circle

dohaduong committed:
Enhanced Duke's personality

dohaduong committed:
Add Personality

dohaduong committed:
Personality enhanced

dohaduong committed:
Resolve conflicts

dohaduong committed:
Resolve conflicts

dohaduong committed:
Resolve conflicts

dohaduong committed:
Merge pull request #7 from dohaduong/A-Personality

A personality

dohaduong committed:
Handle more errors

dohaduong committed:
Update to resolve checkStyle

dohaduong committed:
Improve GUI

dohaduong committed:
Update delete all command

dohaduong committed:
Update Parser test

dohaduong committed:
Added user guide

dohaduong committed:
Update UG

dohaduong committed:
Update according to check style

dohaduong committed:
Added Ui representative

dohaduong committed:
Update README

dohaduong committed:
Update code quality based on feedback from bot

dohaduong committed:
Update delete all command
Delete all command has minor error of clearing the whole task list

Updating delete all command helps users to use it more efficiently

Let's update the command

dohaduong committed:
Delete unused import to improve code quality
